subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultSnapshotHierarchy.java

```diff
@@ -87,11 +88,11 @@ public SnapshotHierarchy empty() {
 
     @Override
     public void visitSnapshots(SnapshotVisitor snapshotVisitor) {
-        rootNode.accept(node -> node.getSnapshot().ifPresent(snapshot -> {
+        rootNode.accept((node, parent) -> node.getSnapshot().ifPresent(snapshot -> {
             if (snapshot instanceof CompleteFileSystemLocationSnapshot) {
-                snapshotVisitor.visitSnapshot((CompleteFileSystemLocationSnapshot) snapshot);
+                snapshotVisitor.visitSnapshot((CompleteFileSystemLocationSnapshot) snapshot, !(parent instanceof CompleteFileSystemLocationSnapshot));
             }
-        }));
+        }), null);
```

Comment:
This looks hard to decipher now.

Reply:
Better now? I could also change the order of the arguments, so the parent comes first and then the visitor. WDYT?